File: charuconstruction/media_reader.py
from abc import ABC, abstractmethod
from typing import Iterable, TYPE_CHECKING

import cv2
import numpy as np

from .camera import Camera
from .frame import Frame
from .math import Transformation, TranslationVector, RotationMatrix, Vector3

if TYPE_CHECKING:
    from .charuco import Charuco


class MediaReader(ABC):
    def __init__(self):
        super().__init__()

    @abstractmethod
    def destroy(self):
        """
        Release any resources held by the reader.
        """

    # Reader as a list iterator
    def __iter__(self) -> "MediaReader":
        return self

    def __next__(self) -> Frame:
        frame = self._read_frame()
        if frame is None:
            raise StopIteration()

        return frame

    @abstractmethod
    def _read_frame(self) -> Frame | None:
        """
        Read a frame from the source.
        """


class ImageReader(MediaReader):
    def __init__(self, image_path: str | Iterable[str]):
        self._image_path = (
            (image_path,) if isinstance(image_path, str) else image_path
        )
        self._index = 0
        super().__init__()

    def destroy(self):
        pass

    def _read_frame(self):
        if self._index >= len(self._image_path):
            return None
        image = cv2.imread(self._image_path[self._index])
        self._index += 1
        return Frame(image)


class VideoReader(MediaReader):
    def __init__(self, video_path: str = 0):
        self._video_path = video_path
        self._cap = cv2.VideoCapture(self._video_path)
        super().__init__()

    def destroy(self):
        self._cap.release()

    def _read_frame(self):
        ret, frame = self._cap.read()
        if not ret:
            return None
        return Frame(frame)


class LiveVideoReader(MediaReader):
    # TODO
    pass


class CharucoMockReader(MediaReader):
    def __init__(
        self,
        boards: Iterable[Charuco],
        camera: Camera,
        transformations: dict[Charuco, Iterable[Transformation]] = None,
    ):
        """
        A mock reader that simulates a video feed of Charuco boards with given transformations.
        If no transformations are provided, a controller window is shown to manually adjust the
        pose of the boards.
        """
        self._boards = boards

        # Sanity checks for angles
        if transformations is None:
            self._frame_count = None
            self._current_index = None
        else:
            self._frame_count = (
                len(transformations[next(iter(transformations))])
                if transformations
                else 0
            )
            self._current_index = 0

            for charuco in transformations.keys():
                if len(transformations[charuco]) != self._frame_count:
                    raise ValueError(
                        "Number of transformations must match number of boards in all frames."
                    )
        self._transformations = transformations

        self._camera = camera
        self._dynamic_frame_ready = False

        super().__init__()

    def destroy(self):
        if self.with_gui:
            for board in self._boards:
                ctrl = f"Controle {board}"
                try:
                    cv2.destroyWindow(ctrl)
                except:
                    pass

    @property
    def with_gui(self) -> bool:
        return self._frame_count is None

    @property
    def frame_count(self) -> int | None:
        return self._frame_count

    def transformations(self, charuco: Charuco) -> Transformation:
        """
        Return the current transformation for a given Charuco board.
        """
        if self.with_gui:
            ctrl = f"Controle {charuco}"
            return Transformation(
                translation=TranslationVector(
                    x=cv2.getTrackbarPos("Trans X", ctrl) / 100,
                    y=cv2.getTrackbarPos("Trans Y", ctrl) / 100,
                    z=cv2.getTrackbarPos("Trans Z", ctrl) / 100,
                ),
                rotation=RotationMatrix.from_euler(
                    Vector3(
                        x=cv2.getTrackbarPos("Rot X", ctrl),
                        y=cv2.getTrackbarPos("Rot Y", ctrl),
                        z=cv2.getTrackbarPos("Rot Z", ctrl),
                    ),
                    sequence=RotationMatrix.Sequence.ZYX,
                    degrees=True,
                ),
            )
        else:
            return self._transformations[charuco][self._current_index - 1]

    def __iter__(self) -> "CharucoMockReader":
        if self.with_gui:
            for board in self._boards:
                ctrl = f"Controle {board}"
                cv2.namedWindow(ctrl, cv2.WINDOW_NORMAL)
                cv2.resizeWindow(ctrl, 400, 300)

                cv2.createTrackbar("Trans X", ctrl, 0, 1000, self._has_moved)
                cv2.setTrackbarMin("Trans X", ctrl, -500)
                cv2.setTrackbarMax("Trans X", ctrl, 500)
                cv2.createTrackbar("Trans Y", ctrl, 0, 1000, self._has_moved)
                cv2.setTrackbarMin("Trans Y", ctrl, -500)
                cv2.setTrackbarMax("Trans Y", ctrl, 500)
                cv2.createTrackbar("Trans Z", ctrl, 150, 500, self._has_moved)
                cv2.setTrackbarMin("Trans Z", ctrl, 0)
                cv2.setTrackbarMax("Trans Z", ctrl, 500)
                cv2.createTrackbar("Rot X", ctrl, 0, 360, self._has_moved)
                cv2.setTrackbarMin("Rot X", ctrl, -360)
                cv2.setTrackbarMax("Rot X", ctrl, 360)
                cv2.createTrackbar("Rot Y", ctrl, 0, 360, self._has_moved)
                cv2.setTrackbarMin("Rot Y", ctrl, -360)
                cv2.setTrackbarMax("Rot Y", ctrl, 360)
                cv2.createTrackbar("Rot Z", ctrl, 0, 360, self._has_moved)
                cv2.setTrackbarMin("Rot Z", ctrl, -360)
                cv2.setTrackbarMax("Rot Z", ctrl, 360)

            self._dynamic_frame_ready = True
        else:
            self._current_index = 0

        return self

    def _has_moved(self, _) -> bool:
        self._dynamic_frame_ready = True

    def _read_frame(self):
        if self.with_gui:
            while not self._dynamic_frame_ready:
                cv2.waitKey(10)
                if cv2.getWindowProperty("frame", cv2.WND_PROP_VISIBLE) == 0:
                    return None
            self._dynamic_frame_ready = False
        else:
            if self._current_index >= self._frame_count:
                return None
            # We increment now since the self.transformations uses the current_index - 1
            self._current_index += 1

        # First create a cv2 image with a white background which corresponds to a
        # distant wall
        frame = np.full(
            (
                int(self._camera.sensor_height),
                int(self._camera.sensor_width),
                3,
            ),
            255,
            dtype=np.uint8,
        )

        # Move the image further away and rotate the boards and get their images
        for board in self._boards:
            img = self._project_board(
                board.cv2_board_image,
                transformation=self.transformations(charuco=board),
            )
            projected_img = cv2.cvtColor(src=img, code=cv2.COLOR_GRAY2BGR)

            masked = projected_img < 255
            frame[masked] = projected_img[masked]

        # Encode and decode to get a proper Frame object
        success, buf = cv2.imencode(".png", frame)
        if not success:
            return None
        return Frame(cv2.imdecode(buf, cv2.IMREAD_COLOR))

    def _project_board(
        self,
        img: np.ndarray,
        transformation: Transformation,
    ) -> np.ndarray:
        """
        Project the Charuco board image with a given rotation and translation.

        Parameters:
            img (np.ndarray): The Charuco board image to project.
            transformation (Transformation): The transformation containing translation and rotation.
        """

        # Corners of the img
        h, w = img.shape[:2]
        s_matrix = np.array(
            [[1, 0, -w / 2], [0, 1, -h / 2], [0, 0, 1]], dtype=np.float32
        )

        translation = (
            transformation.translation.vector * self._camera.focal_length
        )
        rotation = transformation.rotation.matrix
        rt_matrix = np.hstack((rotation, translation))
        h_matrix = self._camera.matrix @ np.delete(rt_matrix, 2, 1)
        h_final = h_matrix @ s_matrix

        return cv2.warpPerspective(
            img,
            h_final,
            (int(self._camera.sensor_width), int(self._camera.sensor_height)),
            flags=cv2.INTER_LINEAR,
            borderValue=255,
        )

File: charuconstruction_flutter/lib/modules/devices/charuco/media_reader.dart
import 'dart:async';
import 'dart:typed_data';

import 'package:camera/camera.dart';
import 'package:logging/logging.dart';
import 'package:ml_linalg/linalg.dart';
import 'package:opencv_dart/opencv.dart';

import 'camera.dart';
import 'charuco.dart';
import 'extensions.dart';
import 'frame.dart';

final _logger = Logger("MediaReader");

abstract class MediaReader {
  ///
  /// Read frames from the media source. This method returns a stream of frames t
  /// hat can be listened to for real-time processing.
  ///
  Stream<Frame?> readFrames();

  ///
  /// Dispose of any resources used by the media reader. This should be called
  /// when the reader is no longer needed to free up memory.
  ///
  void dispose();
}

class ImageReader implements MediaReader {
  final Frame frame;

  ImageReader({required String imagePath}) : frame = Frame(imread(imagePath));

  @override
  Stream<Frame?> readFrames() async* {
    yield frame;
    // Terminate the stream after yielding that single frame
  }

  @override
  void dispose() {
    frame.dispose();
  }
}

class WebcamReader implements MediaReader {
  bool _isReading = false;
  bool get _isInitialized =>
      webcamController != null && webcamController!.value.isInitialized;
  final List<CameraDescription> _availableCameras = [];
  CameraController? webcamController;
  final imageBuffer = <CameraImage>[];

  Future<void> initialize() async {
    _availableCameras.addAll(await availableCameras());
    if (_availableCameras.isEmpty) {
      throw Exception("No cameras available");
    }

    webcamController = CameraController(
      _availableCameras.first,
      ResolutionPreset.high,
      enableAudio: false,
      fps: 300,
    );
    await webcamController!.initialize();
  }

  Future<void> startReading() async {
    if (!_isInitialized) {
      throw Exception("Webcam not initialized. Call initialize() first.");
    }
    if (_isReading) return;

    _isReading = true;
    await webcamController!.startImageStream(_addToBuffer);
  }

  Future<void> stopReading() async {
    if (!_isInitialized) {
      throw Exception("Webcam not initialized. Call initialize() first.");
    }
    if (!_isReading) return;
    await webcamController!.stopImageStream();
    _isReading = false;
  }

  @override
  Stream<Frame?> readFrames() async* {
    while (!_isInitialized) {
      await Future.delayed(Duration(milliseconds: 100));
    }

    while (_isInitialized) {
      try {
        if (imageBuffer.isEmpty) {
          await Future.delayed(Duration(milliseconds: 10));
          continue;
        }
        final cameraImage = imageBuffer.removeAt(0);
        final bytes = _toRgb(frame: cameraImage, rotation: 90);
        if (bytes == null) {
          yield null;
          continue;
        }
        final frame = Mat.fromList(
          cameraImage.width,
          cameraImage.height,
          MatType.CV_8UC(3),
          bytes,
        );
        yield Frame(frame);
      } catch (e) {
        yield null;
      }
    }
    yield null;
  }

  void _addToBuffer(CameraImage bytes) {
    imageBuffer.add(bytes);
    if (imageBuffer.length > 10) {
      _logger.warning(
        "Image buffer overflow: ${imageBuffer.length} frames. Dropping oldest frame.",
      );
      imageBuffer.removeAt(0); // Keep the buffer size manageable
    }
  }

  @override
  void dispose() async {
    await stopReading();
    await webcamController?.dispose();
    webcamController = null;
  }

  Uint8List? _toRgb({required CameraImage frame, int rotation = 90}) {
    try {
      return switch (frame.format.group) {
        ImageFormatGroup.yuv420 => Uint8ListExtension.fromYUV420(
          frame,
          rotation: rotation,
        ),
        ImageFormatGroup.bgra8888 => Uint8ListExtension.fromBGRA8888(
          frame,
          rotation: rotation,
        ),
        _ => throw Exception("Unsupported image format: ${frame.format.group}"),
      };
    } catch (e) {
      _logger.severe("Error converting camera image to RGB: $e");
      return null;
    }
  }
}

class CharucoMockReader implements MediaReader {
  ///
  /// The list of Charuco boards to simulate. Each board will be transformed according to the provided transformations.
  ///
  final List<Charuco> charucos;

  ///
  /// A list of transformations for each Charuco board across all frames.
  /// The outer list corresponds to frames, and the inner list corresponds to the
  /// transformations (translation, rotation) for each board in that frame.
  /// Therefore each inner list should have the same length as the number of Charuco boards.
  ///
  final Stream<List<(Vector, Matrix)>> transformations;

  ///
  /// The camera parameters to use for projecting the Charuco boards in the mock frames.
  /// This includes intrinsic parameters like focal length and sensor size.
  ///
  final Camera camera;

  CharucoMockReader({
    required this.charucos,
    required this.camera,
    required this.transformations,
  });

  @override
  Stream<Frame?> readFrames() async* {
    // Move the board further away and rotate the boards and get their images
    await for (final frameTransformations in transformations) {
      // First create a white background which corresponds to a distant wall
      Mat frame = Mat.fromScalar(
        camera.sensorHeight.toInt(),
        camera.sensorWidth.toInt(),
        MatType.CV_8UC(3),
        Scalar(0xFF / 2, 0xFF / 2, 0xFF / 2),
      );

      for (final (charucoTransformation) in frameTransformations) {
        final translation = charucoTransformation.$1;
        final rotation = charucoTransformation.$2;
        final board =
            charucos[frameTransformations.indexOf(charucoTransformation)];
        final (projectedBoard, mask) = _projectBoard(
          charuco: board,
          camera: camera,
          translation: translation,
          rotation: rotation,
        );

        projectedBoard.copyTo(frame, mask: mask);
      }

      // Encode and decode to get a proper Frame object
      final (isSuccess, buf) = imencode(".png", frame);
      if (!isSuccess) break;
      yield Frame(imdecode(buf, IMREAD_COLOR));
    }

    yield null;
  }

  ///
  /// Project the Charuco board image with a given rotation and translation.
  /// [charuco] The Charuco board to project on the image
  /// [camera] The camera parameters to use for projection.
  /// [frame] The Charuco board image to project.
  /// [transformation] The homogenous transformation (4x4 matrix, consisting of
  /// the 3x3 rotation, 3x1 translation, and 1x1 scale) containing translation
  /// and rotation to apply to the board.
  ///
  /// Returns the projected board image and its mask
  ///
  (Mat, Mat) _projectBoard({
    required Charuco charuco,
    required Camera camera,
    required Vector translation,
    required Matrix rotation,
  }) {
    // Scale the translation part of the transformation by the camera's focal
    // length to simulate perspective projection
    final pixelTranslation = translation * camera.focalLength;

    // Drop the third column of rotation and concatenate the translation
    final projectionMatrix = rotation
        .filterColumns((_, index) => index != 2)
        .insertColumns(2, [pixelTranslation]);

    // Remove the third column of the transformation to get a 3x4 matrix for projection
    final hMatrix = camera.matrixAsLinalg * projectionMatrix;

    // Corners of the img
    final height = charuco.boardImage.rows;
    final width = charuco.boardImage.cols;
    final sMatrix = Matrix.fromList([
      [1, 0, -width / 2],
      [0, 1, -height / 2],
      [0, 0, 1],
    ]);

    // Compute the homography and warp the image
    final hFinal = (hMatrix * sMatrix).toMat();

    // Prepare the mask
    final mask = Mat.fromScalar(height, width, MatType.CV_8UC(1), Scalar(0xFF));

    return (
      warpPerspective(cvtColor(charuco.boardImage, COLOR_GRAY2BGR), hFinal, (
        camera.sensorWidth.toInt(),
        camera.sensorHeight.toInt(),
      )),
      warpPerspective(mask, hFinal, (
        camera.sensorWidth.toInt(),
        camera.sensorHeight.toInt(),
      )),
    );
  }

  @override
  void dispose() {}
}

extension Uint8ListExtension on Uint8List {
  // CameraImage BGRA8888 -> PNG
  // Color
  static Uint8List fromBGRA8888(CameraImage image, {int rotation = 0}) {
    final plane = image.planes[0];

    final width = image.width;
    final height = image.height;
    final bytes = plane.bytes;
    final bytesPerRow = plane.bytesPerRow;

    late Uint8List out;
    late int outWidth;
    late int outHeight;

    if (rotation == 90 || rotation == 270) {
      outWidth = height;
      outHeight = width;
    } else {
      outWidth = width;
      outHeight = height;
    }

    out = Uint8List(outWidth * outHeight * 3);

    for (int y = 0; y < height; y++) {
      final rowStart = y * bytesPerRow;

      for (int x = 0; x < width; x++) {
        final i = rowStart + x * 4;

        final b = bytes[i];
        final g = bytes[i + 1];
        final r = bytes[i + 2];

        int newX, newY;

        switch (rotation) {
          case 90:
            newX = height - y - 1;
            newY = x;
            break;
          case 270:
            newX = y;
            newY = width - x - 1;
            break;
          case 180:
            newX = width - x - 1;
            newY = height - y - 1;
            break;
          default:
            newX = x;
            newY = y;
        }

        final outIndex = (newY * outWidth + newX) * 3;

        out[outIndex] = b;
        out[outIndex + 1] = g;
        out[outIndex + 2] = r;
      }
    }

    return out;
  }

  // CameraImage YUV420_888 -> PNG -> Image (compresion:0, filter: none)
  // Black
  static Uint8List fromYUV420(CameraImage image, {int rotation = 90}) {
    final width = image.width;
    final height = image.height;

    final outWidth = (rotation == 90 || rotation == 270) ? height : width;
    final outHeight = (rotation == 90 || rotation == 270) ? width : height;

    final yPlane = image.planes[0];
    final uPlane = image.planes[1];
    final vPlane = image.planes[2];

    final yBuffer = yPlane.bytes;
    final uBuffer = uPlane.bytes;
    final vBuffer = vPlane.bytes;

    final rgb = Uint8List(outWidth * outHeight * 3);

    final uvRowStride = uPlane.bytesPerRow;
    final uvPixelStride = uPlane.bytesPerPixel!;

    for (int y = 0; y < height; y++) {
      for (int x = 0; x < width; x++) {
        final yIndex = y * yPlane.bytesPerRow + x;
        final uvIndex = (y ~/ 2) * uvRowStride + (x ~/ 2) * uvPixelStride;

        final Y = yBuffer[yIndex];
        final U = uBuffer[uvIndex];
        final V = vBuffer[uvIndex];

        int R = (Y + 1.402 * (V - 128)).round();
        int G = (Y - 0.344136 * (U - 128) - 0.714136 * (V - 128)).round();
        int B = (Y + 1.772 * (U - 128)).round();

        R = R.clamp(0, 255);
        G = G.clamp(0, 255);
        B = B.clamp(0, 255);

        int newX, newY;

        switch (rotation) {
          case 90:
            newX = height - y - 1;
            newY = x;
            break;
          case 180:
            newX = width - x - 1;
            newY = height - y - 1;
            break;
          case 270:
            newX = y;
            newY = width - x - 1;
            break;
          default:
            newX = x;
            newY = y;
        }

        // Use outWidth instead of width
        final newIndex = (newY * outWidth + newX) * 3;

        rgb[newIndex] = R;
        rgb[newIndex + 1] = G;
        rgb[newIndex + 2] = B;
      }
    }

    return rgb;
  }
}
